source/Examples/SmartCache/README.md

```diff
@@ -0,0 +1,123 @@
+
+    ```shell
+    docker build --build-arg=UbuntuVersion=20.04 -t mlos/build:ubuntu-20.04 .
+    ```
+
+2. [Run the Docker image](../../../documentation/02-Build.md#create-a-new-container-instance) you just built.
+
+    ```shell
+    docker run -it -v $PWD:/src/MLOS --name mlos-build mlos/build:ubuntu-20.0
+    ```
+
+3. Inside the container, [build the compiled software](../../../documentation/02-Build.md#cli-make) with `make`:
+
+    ```sh
+    make dotnet-build cmake-build
+    ```
+
+    > This will build everything using a default `CONFIGURATION=Release`.
+    >
+    > To just build `SmartCache` and `Mlos.Agent.Server`, execute the following: \
+    > `make -C source/Examples/SmartCache && make -C source/Mlos.Agent.Server`
+
+4. For a `Release` build, the relevant output will be at:
+
+    - `Mlos.Agent.Server`:
+
+        `out/dotnet/source/Mlos.Agent.Server/obj/AnyCPU/Mlos.Agent.Server.dll`
+
+    - `SmartCache`:
+
+        `out/cmake/Release/source/Examples/SmartCache/SmartCache`
+
+    - `SmartCache.SettingsRegistry`
+
+        `out/dotnet/source/Examples/SmartCache/SmartCache.SettingsRegistry/obj/AnyCPU/SmartCache.SettingsRegistry.dll`
+
+## Executing
+
+`SmartCache` can be invoked separate, or by the `Mlos.Agent.Server` itself.
+
+Once started, `SmartCache` will attempt to register its component specific set of shared memory messages with the `Mlos.Agent` in the `Mlos.Agent.Server` using some `Mlos.Core` component registration messages.  That message includes the name of the `SettingsRegistry` assembly (`.dll`) corresponding to that component's settings/messages.
+
+The `Mlos.Agent.Server` needs to be told where it can find those assemblies.  To do that we provide an `MLOS_SETTINGS_REGISTRY_PATH` environment variable.
+
+In this case we populate it with the path to the `SmartCache.SettingsRegistry.dll`:
+
+```sh
+export MLOS_SETTINGS_REGISTRY_PATH="out/dotnet/source/Examples/SmartCache/SmartCache.SettingsRegistry/obj/AnyCPU:$MLOS_SETTINGS_REGISTRY_PATH"
+```
+
+Next, we can the `Mlos.Server.Agent` using the `dotnet` command:
+
+```sh
+tools/bin/dotnet out/dotnet/source/Mlos.Agent.Server/obj/AnyCPU/Mlos.Agent.Server.dll
+# FIXME: This is missing the .json file to connect to the optimizer service.
+```
+
+The `Mlos.Agent` that gets started will then wait for a signal that the other side (`SmartCache`) has connected to the shared memory region before started to poll it for messages to process.
+
+To start the `SmartCache` process we first need another shell instance in the docker container:
+
+```sh
+docker exec -it mlos-build /bin/bash
+```
+
+Now, we can start `SmartCache` as follows:
+
+```sh
+out/cmake/Release/source/Examples/SmartCache/SmartCache
+```
+
+> To have `Mlos.Agent.Server` start `SmartCache` without having to start another shell in the docker container instance, add the path to the `SmartCache` binary as an argument to the `dotnet ... Mlos.Agent.Server` invocation.
```

Comment:
wouldn't it be easier / shorter to just give that instruction? I guess that doesn't show off the separation of the executable and the agent that well. Maybe removing that comment for the default instructions will make it easier? not sure.

Reply:
Ack.  I think I was struggling with how to show the two methods, and maybe that's not necessary here.  Perhaps it'd be better suited in a `README.md` for `Mlos.Agent.Server`.

What's your preference when driving this from a notebook?

1. Have a single invocation that handles both processes (e.g. `dotnet path/to/Mlos.Agent.Server.dll path/to/SmartCache), or
2. Have two invocations:
  1. To start Mlos.Agent.Server
  2. To (repeatedly) start SmartCache to submit new results to it.

Reply:
I think the second is more conceptually clear?

Reply:
Ok, I did the first just now, but I can change it back.

Reply:
either is fine :)

Reply:
Cool, let's go with this for the moment.  We can change it later.